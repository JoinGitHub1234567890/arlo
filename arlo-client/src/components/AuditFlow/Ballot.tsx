import React, { useState } from 'react'
import { H1, H3, Callout, H4, Divider } from '@blueprintjs/core'
import styled from 'styled-components'
import BallotAudit from './BallotAudit'
import BallotReview from './BallotReview'
import { AuditBoard } from '../../types'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`

const MainCallout = styled(Callout)`
  width: 400px;
  font-weight: 700;
`

const Info = styled.div`
  display: flex;
  justify-content: flex-start;
  margin: 20px 0;

  .ballot-info {
    width: 200px;
    padding: 20px 0;
  }
  .description {
    width: 50%;
    padding: 20px;
  }
`

interface Props {
  roundId: string
  ballotId: string
  board: AuditBoard
}

const Ballot: React.FC<Props> = ({ roundId, ballotId, board }: Props) => {
  const [auditing, setAuditing] = useState(true)

  if (!board.ballots) return null // TODO handle informatively
  const ballot = board.ballots[Number(ballotId) - 1]
  if (!ballot) return null // TODO handle informatively

  return (
    <Wrapper>
      <H1>{board.name}: Ballot Card Data Entry</H1>
      <H3>Enter Ballot Information</H3>
      <MainCallout icon={null} intent="primary">
        Round {roundId}: auditing ballot {ballotId} of {board.ballots.length}
      </MainCallout>
      <Info>
        <div className="ballot-info">
          <H4>Current ballot:</H4>
          <div>Tabulator: {ballot.tabulator}</div>
          <div>Batch: {ballot.batch}</div>
          <div>Record/Position: {ballot.record}</div>
        </div>
        <Divider />
        <div className="description">
          <H4>Are you looking at the correct ballot?</H4>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in
            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
            pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
            culpa qui officia deserunt mollit anim id est laborum.
          </p>
        </div>
      </Info>
      {auditing ? (
        <BallotAudit review={() => setAuditing(false)} />
      ) : (
        <BallotReview audit={() => setAuditing(true)} />
      )}
    </Wrapper>
  )
}

export default Ballot
