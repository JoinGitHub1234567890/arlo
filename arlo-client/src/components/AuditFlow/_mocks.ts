import { IAuditBoard } from '../../types'

export const dummyBoard: IAuditBoard[] = [
  {
    id: '123',
    name: 'Audit Board #1',
    members: [],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: 'Precinct 13',
        position: '313',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: 'Precinct 13',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '17',
        batch: 'Precinct 19',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
      {
        tabulator: '23',
        batch: 'Precinct 29',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
      },
    ],
  },
]

export default {
  dummyBoard,
}
