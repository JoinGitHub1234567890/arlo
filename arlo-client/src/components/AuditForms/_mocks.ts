import { Audit } from '../../types'

export const mockAudit: Audit = {
  name: 'Primary 2019',
  riskLimit: 10,
  randomSeed: 123123123,

  contests: [
    {
      id: 'contest-1',
      name: 'Contest 1',

      choices: [
        {
          id: 'candidate-1',
          name: 'Candidate 1',
          numVotes: 42,
        },
        {
          id: 'candidate-2',
          name: 'Candidate 2',
          numVotes: 72,
        },
      ],

      totalBallotsCast: 4200,
    },
  ],

  jurisdictions: [
    {
      id: 'adams-county',
      name: 'Adams County',
      contests: ['contest-1'],
      auditBoards: [
        {
          id: 'audit-board-1',
          members: [],
        },
        {
          id: 'audit-board-2',
          members: [],
        },
      ],
      ballotManifest: {
        filename: 'Adams_County_Manifest.csv',
        numBallots: 123456,
        numBatches: 560,
        uploadedAt: '2019-06-17 11:45:00',
      },
    },
  ],

  rounds: [
    {
      startedAt: '2019-06-17 11:45:00',
      endedAt: '2019-06-17 11:55:00',
      contests: [
        {
          id: 'contest-1',
          endMeasurements: {
            pvalue: 0.085,
            isComplete: false,
          },
          results: {
            'candidate-1': 55,
            'candidate-2': 35,
          },
          sampleSize: 25,
        },
      ],
      jurisdictions: {
        'adams-county': {
          numBallots: 15,
        },
      },
    },
  ],
}

export default mockAudit
