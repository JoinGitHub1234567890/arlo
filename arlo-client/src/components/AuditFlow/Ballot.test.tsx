import React from 'react'
import { render, fireEvent, wait, waitForElement } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import Ballot from './Ballot'
import { dummyBallots } from './_mocks'

const history = createMemoryHistory()

describe('Ballot', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={jest.fn()}
          nextBallot={jest.fn()}
          submitBallot={jest.fn()}
          roundIx="1"
          batchId="batch-id-1"
          ballotId={313}
        />
      </Router>
    )
    expect(container).toMatchSnapshot()
  })

  it('switches audit and review views', async () => {
    const { container, getByText, getByTestId } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={jest.fn()}
          nextBallot={jest.fn()}
          submitBallot={jest.fn()}
          roundIx="1"
          batchId="batch-id-1"
          ballotId={313}
        />
      </Router>
    )

    fireEvent.click(getByTestId('YES'), { bubbles: true })
    await wait(() =>
      fireEvent.click(getByTestId('enabled-review'), { bubbles: true })
    )
    await wait(() => {
      expect(getByText('Submit & Next Ballot')).toBeTruthy()
    })
    await wait(() => {
      expect(container).toMatchSnapshot()
    })
    fireEvent.click(getByText('Edit'), { bubbles: true })
    await wait(() => {
      expect(getByText('Review')).toBeTruthy()
    })
  })

  it('toggles and submits comment', async () => {
    const { container, getByText, getByTestId } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={jest.fn()}
          nextBallot={jest.fn()}
          submitBallot={jest.fn()}
          roundIx="1"
          batchId="batch-id-1"
          ballotId={313}
        />
      </Router>
    )

    const commentInput = getByTestId('comment-textarea')
    fireEvent.change(commentInput, { target: { value: 'a test comment' } })

    fireEvent.click(getByTestId('YES'), { bubbles: true })
    await wait(() =>
      fireEvent.click(getByTestId('enabled-review'), { bubbles: true })
    )
    await wait(() => {
      expect(getByText('Submit & Next Ballot')).toBeTruthy()
      expect(getByText('COMMENT: a test comment')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })

  it('submits review and progresses to next ballot', async () => {
    const submitMock = jest.fn()
    const nextBallotMock = jest.fn()
    const { container, getByText, getByTestId } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={jest.fn()}
          nextBallot={nextBallotMock}
          submitBallot={submitMock}
          roundIx="1"
          batchId="batch-id-1"
          ballotId={313}
        />
      </Router>
    )

    fireEvent.click(getByTestId('YES'), { bubbles: true })

    const reviewButton = await waitForElement(
      () => getByTestId('enabled-review'),
      { container }
    )
    fireEvent.click(reviewButton, { bubbles: true })
    const nextButton = await waitForElement(
      () => getByText('Submit & Next Ballot'),
      { container }
    )
    fireEvent.click(nextButton, { bubbles: true })

    await wait(() => {
      expect(nextBallotMock).toBeCalled()
      expect(submitMock).toHaveBeenCalledTimes(1)
    })
  })

  it('navigates to previous ballot', async () => {
    const previousBallotMock = jest.fn()
    const { getByText, getByTestId } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={previousBallotMock}
          nextBallot={jest.fn()}
          submitBallot={jest.fn()}
          roundIx="1"
          batchId="batch-id-1"
          ballotId={2112}
        />
      </Router>
    )
    fireEvent.click(getByText('Back'), { bubbles: true })

    await wait(() => {
      expect(previousBallotMock).toBeCalledTimes(1)
    })

    fireEvent.click(getByTestId('YES'), { bubbles: true })
    await wait(() =>
      fireEvent.click(getByTestId('enabled-review'), { bubbles: true })
    )
    await wait(() => {
      expect(getByText('Submit & Next Ballot')).toBeTruthy()
    })
    fireEvent.click(getByText('Back'), { bubbles: true })
    await wait(() => {
      expect(previousBallotMock).toBeCalledTimes(2)
    })
  })

  it('redirects if ballot does not exist', async () => {
    const { container } = render(
      <Router history={history}>
        <Ballot
          home="/election/1/board/1"
          ballots={dummyBallots.ballots}
          boardName="audit board #1"
          contest="contest name"
          previousBallot={jest.fn()}
          nextBallot={jest.fn()}
          submitBallot={jest.fn()}
          roundIx="1"
          batchId="batch-id"
          ballotId={6}
        />
      </Router>
    )

    await wait(() => {
      expect(container).toMatchSnapshot()
    })
  })
})
