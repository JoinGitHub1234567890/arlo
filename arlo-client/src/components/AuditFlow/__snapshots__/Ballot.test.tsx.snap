// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ballot redirects if ballot does not exist 1`] = `<div />`;

exports[`Ballot renders correctly 1`] = `
<div>
  <div
    class="sc-fjdhpX kEvzkS"
  >
    <h1
      class="bp3-heading sc-jTzLTM eHLtYK"
    >
      audit board #1
      : Ballot Card Data Entry
    </h1>
    <h3
      class="bp3-heading"
    >
      Enter Ballot Information
    </h3>
    <div
      class="bp3-callout sc-jzJRlG iQxVZM"
    >
      Round 
      1
      : auditing ballot 
      1
       of 
      27
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      >
        <h4
          class="bp3-heading"
        >
          Current ballot:
        </h4>
        <div>
          Tabulator: 
        </div>
        <div>
          Batch: 
          0003-04-Precinct 13 (Jonesboro Fire Department)
        </div>
        <div>
          Record/Position: 
          313
        </div>
      </div>
      <div
        class="bp3-divider"
      />
      <div
        class="ballot-main"
      >
        <h4
          class="bp3-heading"
        >
          Are you looking at the correct ballot?
        </h4>
        <p>
          Before continuing, check the "Current ballot" information to make sure you are entering data for the correct ballot. If the ballot could not be found, click "Ballot not found" below and move on to the next ballot.
        </p>
        <p>
          <button
            class="bp3-button bp3-intent-danger"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Ballot 
              313
               not found - move to next ballot
            </span>
          </button>
        </p>
        <p>
          <a
            class="bp3-button bp3-intent-primary"
            href="/election/1/board/1"
          >
            Return to audit overview
          </a>
        </p>
      </div>
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      />
      <div
        class="ballot-main"
      >
        <h4
          class="bp3-heading"
        >
          Instructions
        </h4>
        <p>
          Select 
          <strong>
            all
          </strong>
           the candidates/choices below that you see marked on the paper ballot.
        </p>
        <p>
          If the voter did not vote in the contest, select "Blank vote/no mark."
        </p>
        <p>
          If the audit board cannot agree, select "No audit board consensus." You may add a comment for additional information about the disagreement.
        </p>
        <form
          class="sc-bwzfXH imPamb"
        >
          <h3
            class="bp3-heading"
          >
            contest name
          </h3>
          <div
            class="bp3-divider"
          />
          <div
            class="sc-htpNat fisUI"
          >
            <label
              class="undefined bp3-control bp3-radio sc-EHOje cxPscF"
            >
              <input
                data-testid="YES"
                name="vote"
                type="radio"
                value="YES"
              />
              <span
                class="bp3-control-indicator"
              >
                <span
                  class="radio-text"
                >
                  Yes/For
                </span>
              </span>
            </label>
            <label
              class="undefined bp3-control bp3-radio sc-EHOje cxPscF"
            >
              <input
                data-testid="NO"
                name="vote"
                type="radio"
                value="NO"
              />
              <span
                class="bp3-control-indicator"
              >
                <span
                  class="radio-text"
                >
                  No/Against
                </span>
              </span>
            </label>
            <label
              class="undefined bp3-control bp3-radio sc-EHOje cxPscF"
            >
              <input
                data-testid="NO_CONSENSUS"
                name="vote"
                type="radio"
                value="NO_CONSENSUS"
              />
              <span
                class="bp3-control-indicator"
              >
                <span
                  class="radio-text"
                >
                  No audit board consensus
                </span>
              </span>
            </label>
            <label
              class="undefined bp3-control bp3-radio sc-EHOje cxPscF"
            >
              <input
                data-testid="NO_VOTE"
                name="vote"
                type="radio"
                value="NO_VOTE"
              />
              <span
                class="bp3-control-indicator"
              >
                <span
                  class="radio-text"
                >
                  Blank vote/no mark
                </span>
              </span>
            </label>
          </div>
          <div
            class="sc-gzVnrw lfdfns"
          >
            <textarea
              class="bp3-input sc-dnqmqq nIqCe"
              data-testid="comment-textarea"
              name="comment"
            />
          </div>
        </form>
        <div
          class="sc-bxivhb dewsOl"
        >
          <button
            class="bp3-button bp3-disabled bp3-intent-success sc-ifAKCX kKfkZO"
            data-testid="disabled-review"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              Review
            </span>
          </button>
          <button
            class="bp3-button bp3-minimal"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Back
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Ballot switches audit and review views 1`] = `
<div>
  <div
    class="sc-fjdhpX kEvzkS"
  >
    <h1
      class="bp3-heading sc-jTzLTM eHLtYK"
    >
      audit board #1
      : Ballot Card Data Entry
    </h1>
    <h3
      class="bp3-heading"
    >
      Enter Ballot Information
    </h3>
    <div
      class="bp3-callout sc-jzJRlG iQxVZM"
    >
      Round 
      1
      : auditing ballot 
      1
       of 
      27
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      >
        <h4
          class="bp3-heading"
        >
          Current ballot:
        </h4>
        <div>
          Tabulator: 
        </div>
        <div>
          Batch: 
          0003-04-Precinct 13 (Jonesboro Fire Department)
        </div>
        <div>
          Record/Position: 
          313
        </div>
      </div>
      <div
        class="bp3-divider"
      />
      <div
        class="ballot-main"
      >
        <h4
          class="bp3-heading"
        >
          Are you looking at the correct ballot?
        </h4>
        <p>
          Before continuing, check the "Current ballot" information to make sure you are entering data for the correct ballot. If the ballot could not be found, click "Ballot not found" below and move on to the next ballot.
        </p>
        <p>
          <button
            class="bp3-button bp3-intent-danger"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Ballot 
              313
               not found - move to next ballot
            </span>
          </button>
        </p>
        <p>
          <a
            class="bp3-button bp3-intent-primary"
            href="/election/1/board/1"
          >
            Return to audit overview
          </a>
        </p>
      </div>
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      />
      <div
        class="ballot-main"
      >
        <form
          class="sc-bwzfXH imPamb"
        >
          <h3
            class="bp3-heading"
          >
            contest name
          </h3>
          <div
            class="bp3-divider"
          />
          <div
            class="sc-gqjmRU eYlscI"
          >
            <div
              class="bp3-button-group bp3-fill bp3-large bp3-vertical"
            >
              <button
                class="bp3-button bp3-disabled bp3-fill bp3-large bp3-intent-primary sc-bZQynM hJohEA sc-ifAKCX kKfkZO"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  Yes/For
                </span>
              </button>
              <button
                class="bp3-button"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  Edit
                </span>
              </button>
            </div>
          </div>
          <p>
            COMMENT: 
            
          </p>
        </form>
        <div
          class="sc-bxivhb dewsOl"
        >
          <button
            class="bp3-button bp3-intent-success sc-ifAKCX kKfkZO"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              Submit & Next Ballot
            </span>
          </button>
          <button
            class="bp3-button bp3-minimal"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Back
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Ballot toggles and submits comment 1`] = `
<div>
  <div
    class="sc-fjdhpX kEvzkS"
  >
    <h1
      class="bp3-heading sc-jTzLTM eHLtYK"
    >
      audit board #1
      : Ballot Card Data Entry
    </h1>
    <h3
      class="bp3-heading"
    >
      Enter Ballot Information
    </h3>
    <div
      class="bp3-callout sc-jzJRlG iQxVZM"
    >
      Round 
      1
      : auditing ballot 
      1
       of 
      27
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      >
        <h4
          class="bp3-heading"
        >
          Current ballot:
        </h4>
        <div>
          Tabulator: 
        </div>
        <div>
          Batch: 
          0003-04-Precinct 13 (Jonesboro Fire Department)
        </div>
        <div>
          Record/Position: 
          313
        </div>
      </div>
      <div
        class="bp3-divider"
      />
      <div
        class="ballot-main"
      >
        <h4
          class="bp3-heading"
        >
          Are you looking at the correct ballot?
        </h4>
        <p>
          Before continuing, check the "Current ballot" information to make sure you are entering data for the correct ballot. If the ballot could not be found, click "Ballot not found" below and move on to the next ballot.
        </p>
        <p>
          <button
            class="bp3-button bp3-intent-danger"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Ballot 
              313
               not found - move to next ballot
            </span>
          </button>
        </p>
        <p>
          <a
            class="bp3-button bp3-intent-primary"
            href="/election/1/board/1"
          >
            Return to audit overview
          </a>
        </p>
      </div>
    </div>
    <div
      class="sc-bdVaJa kMtymi"
    >
      <div
        class="ballot-side"
      />
      <div
        class="ballot-main"
      >
        <form
          class="sc-bwzfXH imPamb"
        >
          <h3
            class="bp3-heading"
          >
            contest name
          </h3>
          <div
            class="bp3-divider"
          />
          <div
            class="sc-gqjmRU eYlscI"
          >
            <div
              class="bp3-button-group bp3-fill bp3-large bp3-vertical"
            >
              <button
                class="bp3-button bp3-disabled bp3-fill bp3-large bp3-intent-primary sc-bZQynM hJohEA sc-ifAKCX kKfkZO"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  Yes/For
                </span>
              </button>
              <button
                class="bp3-button"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  Edit
                </span>
              </button>
            </div>
          </div>
          <p>
            COMMENT: 
            a test comment
          </p>
        </form>
        <div
          class="sc-bxivhb dewsOl"
        >
          <button
            class="bp3-button bp3-intent-success sc-ifAKCX kKfkZO"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              Submit & Next Ballot
            </span>
          </button>
          <button
            class="bp3-button bp3-minimal"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              Back
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
